### PR TITLE
consul_kv: migrate to shared _consul module utility

### PR DESCRIPTION
**Pull request by PjSalty**
##### SUMMARY

Migrate `consul_kv` to the shared `community.general._consul` module utility
and documentation fragments, per the direction in #12216: get `consul_kv`
closer to the shared util before adding features to individual modules.

First commit extends the shared `_ConsulModule` request layer so KV fits:
bytes/bytearray request bodies pass through without JSON encoding (KV values
are opaque bytes), a new `request()` method returns response headers alongside
the parsed body (KV needs `X-Consul-Index`), and `RequestError` carries
response headers so a 404 handler can still read the index. The existing
`get`/`put`/`delete` wrappers keep their body-only return shape, so the nine
modules already on the shared util need no changes.

Second commit rewires `consul_kv` on top of that: drops the `py-consul` and
`requests` runtime dependencies, uses `AUTH_ARGUMENTS_SPEC` plus the
`_consul`/`_consul.token` doc fragments for connection options, and joins the
`community.general.consul` action group. KV-specific concerns stay in the
module: base64-decoding `Value`, the recurse list-versus-single result shape,
translating 404 to the `(index, None)` shape py-consul produced, and
percent-encoding keys (py-consul rode on `requests`, which quotes URL paths -
keys containing spaces or non-ASCII are legal in Consul and must keep
working).

Behaviour, return shape, and the existing option set are unchanged.

`consul.py` is the remaining module off the shared util; it has its own
py-consul >= 1.4 breakage tracked in #11040 and is left for a follow-up.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
consul_kv

##### ADDITIONAL INFORMATION

Verified locally against `hashicorp/consul` 1.20, 1.22.7 (the version CI
uses) and 2.0.0, each with ACLs enabled:

- side-by-side parity run against the previous py-consul-based module:
  byte-identical results across 16 KV operations (PUT/GET/DELETE, recurse,
  CAS, unicode values, idempotent updates, nested paths, empty-string
  values, 404 handling)
- the `tests/integration/targets/consul` target via
  `ansible-test integration --docker default consul`
- an edge case matrix: CAS create-only/current/stale, session lock
  acquire/release, ACL denial, deep slash paths, multi-line values,
  check mode, 100KB values, keys containing spaces and non-ASCII
- TLS with CA validation, with `validate_certs=false`, and the
  fail-on-self-signed default
- `module_defaults: group/community.general.consul` now applies to
  `consul_kv`
- `consul_session` and `consul_policy` against the updated shared util
- `ansible-test sanity --docker default` (Python 3.9-3.14)

### CODE REVIEW

File: changelogs/fragments/consul-module-utils-bytes-and-headers.yml
Comment (russoz) [inline changelogs/fragments/consul-module-utils-bytes-and-headers.yml:1]: The changelogs:
- These changelog are quite verbose, I would recommend making them more concise. There are many low-level details that are unlikely going to matter for users - many of them are not programmers.
- Separate files _works_, but I would recommend using one single file for the two entries
- Also please note that you must add the URL for the issue, when applicable, and for the PR.

Please refer to the standard formatting for the changelog fragments' entries:

https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to-format

The outcome of what you write will be published in the collection changelog:

https://docs.ansible.com/ansible/latest/collections/community/general/changelog.html

File: plugins/modules/consul_kv.py [hunk 1]
```diff
@@ -142,35 +114,76 @@
...
+    # (spaces, non-ASCII, ...). Slashes are key separators and must stay raw.
+    return quote(key, safe="/")
+
+
+def _kv_get(consul_module, key, recurse=False, dc=None):
+    params = {"dc": dc}
+    if recurse:
+        params["recurse"] = "true"
+    try:
+        body, headers = consul_module.request("GET", ("kv", _quote_key(key)), params=params)
```
Comment (russoz) [inline plugins/modules/consul_kv.py:147]: Why not use `consul_module.get()`?

File: plugins/modules/consul_kv.py [hunk 2]
```diff
@@ -142,35 +114,76 @@
...
+    RequestError,
+    _ConsulModule,
+)
 
-    python_consul_installed = True
-except ImportError:
-    python_consul_installed = False
+# Sentinel marking "the caller did not pass a value" - distinct from None/empty
+# string, which are legitimate states for a KV entry.
+NOT_SET = None
```
Comment (russoz) [inline plugins/modules/consul_kv.py]: This comment makes no sense - this constant is used to check whether a parameters has been passed to the module. When params are not passed, they carry the value `None` which is exactly the value of this constant. This is _not distinct from `None`.

File: plugins/modules/consul_kv.py [hunk 3]
```diff
@@ -181,145 +194,128 @@ def _has_value_changed(consul_client, key, target_value):
...
     elif state == "present":
         if module.params.get("value") is NOT_SET:
-            get_value(module)
+            get_value(module, consul_module)
         else:
-            set_value(module)
+            set_value(module, consul_module)
     elif state == "absent":
-        remove_value(module)
+        remove_value(module, consul_module)
```
Comment (russoz) [inline plugins/modules/consul_kv.py:200]: This bothers me a little bit: the code already had, IMHO, a poor structure, passing only the AnsibleModule object to every action. Now we also pass the same object to every action as well.

You might want to consider using `ModuleHelper`. See more in https://docs.ansible.com/ansible/latest/collections/community/general/docsite/guide_modulehelper.html

File: plugins/modules/consul_kv.py [hunk 4]
```diff
@@ -181,145 +194,128 @@ def _has_value_changed(consul_client, key, target_value):
...
-    module.exit_json(changed=False, index=index, data=existing_value)
-
-
-def set_value(module):
-    consul_api = get_consul_api(module)
 
+def set_value(module, consul_module):
     key = module.params.get("key")
     value = module.params.get("value")
+    dc = module.params.get("datacenter")
```
Comment (russoz) [inline plugins/modules/consul_kv.py]: I know the rest of the code was already like that, but please change these lines to:
```suggestion
    dc = module.params["datacenter"]
```

Ansible **always** populate the `.params` dict with all parameters (passed and otherwise).

**Review by russoz — COMMENTED** [review]
hi @PjSalty 

Thanks for your contribution!

I got a couple of comments.

File: plugins/modules/consul_kv.py (hunk 2 above)
Comment (felixfontein) [reply plugins/modules/consul_kv.py]: A proper sentinel would be `NOT_SET = object()`.

File: changelogs/fragments/consul-module-utils-bytes-and-headers.yml
Comment (PjSalty) [reply changelogs/fragments/consul-module-utils-bytes-and-headers.yml:1]: Merged the two into one file and trimmed it to the user-facing bits (no more py-consul dep, action group) with the PR link. Didn't add one for the `_consul.py` change since it's a private util and doesn't change any existing module's behavior, but happy to add one if you'd rather.

File: plugins/modules/consul_kv.py (hunk 1 above)
Comment (PjSalty) [reply plugins/modules/consul_kv.py:147]: `get()` only returns the body and drops the response headers. The read needs `X-Consul-Index` off the response for the `index` return, so it goes through `request()`. Put and delete don't need it.

File: plugins/modules/consul_kv.py (hunk 2 above)
Comment (PjSalty) [reply plugins/modules/consul_kv.py]: That comment was wrong, `NOT_SET` was just `None`. `value` only has two states (not passed is read, passed is write), and `None` already covers not-passed for a str param while `""` still writes. Dropped the name and the default, the check is `value is None` now.

Comment (PjSalty) [reply plugins/modules/consul_kv.py]: I tried `object()` for this: on a `type="str"` param ansible coerces the default to the string `'<object object at 0x...>'` so the identity check breaks. It only holds up under `type="raw"`, which loses the str validation, so I went with `value is None` instead.

File: plugins/modules/consul_kv.py (hunk 4 above)
Comment (PjSalty) [reply plugins/modules/consul_kv.py]: Switched these and the rest to `module.params[...]`.

File: plugins/modules/consul_kv.py (hunk 3 above)
Comment (PjSalty) [reply plugins/modules/consul_kv.py:200]: The threading's the same as consul_session. Cleanest fix for consistency is probably subclassing `_ConsulModule` and overriding `execute()`, like the 7 modules that already subclass it (policy, token, role, auth_method, binding_rule, agent_check, agent_service). KV doesn't fit the base CRUD `execute()` though (raw byte values, the acquire/release states it rejects, recurse), so it'd reuse the HTTP methods and override `execute()`. I prototyped that and it comes out parity-clean.

Other options are composition (hold a `_ConsulModule` instead of subclassing) or ModuleHelper, though MH would make consul_kv the only one in the family using it. Didn't want to pick the structure for you, so which way do you want to go?

File: plugins/modules/consul_kv.py [hunk 5]
```diff
@@ -16,19 +16,21 @@
...
+  - community.general._consul.actiongroup_consul
+  - community.general._consul.token
   - community.general._attributes
 attributes:
   check_mode:
     support: full
   diff_mode:
     support: none
+  action_group:
+    version_added: 13.1.0
```
Comment (russoz) [inline plugins/modules/consul_kv.py]:
```suggestion
    version_added: 13.2.0
```

File: plugins/modules/consul_kv.py [hunk 6]
```diff
@@ -142,35 +114,72 @@
...
+    if recurse:
+        return index, body
+    return index, body[0]
+
+
+def _kv_put(consul_module, key, value, cas=None, acquire=None, release=None, flags=None, dc=None):
+    params = {"cas": cas, "acquire": acquire, "release": release, "flags": flags, "dc": dc}
+    if value is None:
+        body = b""
+    elif isinstance(value, str):
```
Comment (russoz) [inline plugins/modules/consul_kv.py]: That problem is a low-level one and it should not be dealt with at this level. The code around the HTTP request should handle that kind of situation - and in fact it does. This sanitisation is redundant.
```suggestion
    if isinstance(value, str):
```

**Review by russoz — COMMENTED** [review]
Hi @PjSalty 

Thanks for your continued effort. A couple of comments more.

File: plugins/modules/consul_kv.py (hunk 5 above)
Comment (PjSalty) [reply plugins/modules/consul_kv.py]: Bumped to 13.2.0.

File: plugins/modules/consul_kv.py (hunk 6 above)
Comment (PjSalty) [reply plugins/modules/consul_kv.py]: Right, dropped it. `value` is only ever str or None here (it's a `type="str"` param), and a None body is an empty PUT same as `b""`, so it's just the str-to-bytes encode now. Verified the empty-value write and lock-acquire-with-no-value paths still behave the same against a live Consul.

File: plugins/module_utils/_consul.py [hunk 7]
```diff
@@ -298,8 +299,9 @@ def _request(self, method, url_parts, data=None, params=None):
         if token:
             headers["X-Consul-Token"] = token
 
+        response_headers = None
         try:
-            if data is not None:
+            if data is not None and not isinstance(data, (bytes, bytearray)):
```
Comment (felixfontein) [inline plugins/module_utils/_consul.py]: Why `bytearray`? I don't see that used anywhere.

Comment (PjSalty) [reply plugins/module_utils/_consul.py]: Yeah, habit. `(bytes, bytearray)` is just my reflex check for an already-encoded body, but nothing here passes a bytearray, so good catch. Narrowed to `bytes`.

File: plugins/modules/consul_kv.py [hunk 8]
```diff
@@ -142,35 +114,68 @@
...
+        _decode_value(entry)
+    if recurse:
+        return index, body
+    return index, body[0]
+
+
+def _kv_put(consul_module, key, value, cas=None, acquire=None, release=None, flags=None, dc=None):
+    params = {"cas": cas, "acquire": acquire, "release": release, "flags": flags, "dc": dc}
+    if isinstance(value, str):
+        value = value.encode("utf-8")
```
Comment (felixfontein) [inline plugins/modules/consul_kv.py:166]: The way I can see this , `value` is only ever `None` or a string. Why not check `if value is not None` instead?

**Review by felixfontein — COMMENTED** [review]
Generally when attempting such a large rewrite, it's best to use type hints for all new code - this makes the code and PR both easier to understand, and allows our tooling to check for more errors.

In any case, the PR looks solid, and if nobody objects I'll merge it for the next release (~Monday).

File: plugins/modules/consul_kv.py (hunk 8 above)
Comment (PjSalty) [reply plugins/modules/consul_kv.py:166]: Yeah it's only ever None or a string here, the None being a lock acquire/release without a value where the PUT has to go out with an empty body. Changed it.

**Comment by PjSalty** [comment]
Noted on the type hints, I'll bring them into the consul follow-up work rather than churn this one before the merge. Thanks for the review.

**Comment by felixfontein** [comment]
Thanks for fixing this, LGTM from my side!

Another thing for later: I'd remove the leading underscores for (most) function names in `consul_kv.py`. Everything in that file is private anyway.

**Comment by felixfontein** [comment]
@PjSalty thanks for your contribution!
@russoz thanks for reviewing!
